### PR TITLE
Setup new baseline with ERA5 forcing and new global land grids 

### PR DESCRIPTION
**Pull request by rhaesung**
## Description
1. The latest ufs-land-driver (https://github.com/NOAA-EMC/ufs-land-driver.git@203236b) is updated with CMake feature. The main updates are:

- added mpi capability
- added monthly, daily, solar noon output capability   

2. The latest vector2tile (https://github.com/NOAA-PSL/land-vector2tile.git@2fb5c1b) is loaded. The main update is:

- Changed snow_layers to active_snow_layers for new restarts

3. New baseline is created using the new date of GHCN, ERA5 forcing and the new global land grid points.

- updated observations path
- updated initial condition, restart file, forcing directory pathes

4. In the case of 'hera.internal' baseline,

- The latest JEDI internal version (https://github.com/JCSDA-internal/fv3-bundle.git@bd6acdc) utilized. Need to merge https://github.com/ufs-community/land-DA/pull/2
- Required modules are loaded.

5. Added PSL's ICS checking option and enable to write the first DA day's background output.

6. 'hera.internal' baseline option is tested and validated with the current PSL develop for 2019-12-21 ~2019-12-23. All DA results are identical (PSL DA outputs: /scratch1/NCEPDEV/stmp2/Rhaesung.Kim/psl_develop_062223/land-offline_workflow/cycle_land/DA_test_era5)

7. ctest is functional

8. Input datasets and test_base_directory for ctest are relocated under NAGAPE/epic on hera. Needed the same work for other platforms.

### Anticipated changes to regression tests:
- [X] Is baseline change expected ? 

## Subcomponents involved:
- [X] DA_update
- [X] vector2tile
- [X] ufs-land-driver
- [ ] none

## Linked PR's and Issues:
https://github.com/ufs-community/land-DA_workflow/issues/21
https://github.com/ufs-community/land-DA_workflow/issues/22

### Testing (for CM's):
- RDHPCS
    - [X] Hera
    - [ ] Orion
    - [ ] Jet
    - [ ] Gaea
    - [ ] Cheyenne
- CI
  - [ ] Completed
- PW-Clouds
  - [ ] AWS
  - [ ] AZURE
  - [ ] GCP

- Tips for building and testing steps:
cd vector2tile ->./configure-> make
cd ufs-land-driver -> ./configure -> make
cd DA_update -> ./build_all.sh
./do_submit_cycle.sh settings_DA_cycle_era5

### CODE REVIEW

**Comment by jkbk2004** [comment]
@rhaesung Its good to see it reproduces PSL develop. What about ctests? Just wondering about any impact on cmake side. can you show ctest results with your branch?

**Comment by jkbk2004** [comment]
@rhaesung You mentioned you have tested with skylab4.0. Can you create an issue to follow up skylab4.0: installation steps and any tests you did.

**Comment by jkbk2004** [comment]
@rhaesung regarding sanity check for cmake/ctest, can you run ctests in singularity container? Do you think some additional lines in readme file would be helpful? then please, go ahead to add some new build and test information in readme or changelog files.

**Comment by rhaesung** [comment]
> @rhaesung You mentioned you have tested with skylab4.0. Can you create an issue to follow up skylab4.0: installation steps and any tests you did.

I did a test with skylab4.0 using the non-hera.internal baseline.  I will create an issue for this with the installation steps.

**Comment by rhaesung** [comment]
@jkbk2004 Sure, I will work on testing cmake/ctest and check if there is anything I can update in the readme.

**Comment by rhaesung** [comment]
@jkbk2004 @zach1221 I was able to create executables using ‘cmake’. ufs-land-driver@203236b is missing CMakeLists.txt and the new CMakeLists.txt is needed and needs to be updated pointing to the ccpp-physics@df5f5e2. DA results are not compared yet.

Most ctests are currently failing because of a new baseline with new inputs. I will go ahead and try if I can fix this, or probably needs to be coordinated with @clouden90.

**Comment by jkbk2004** [comment]
@rhaesung are you testing under singularity? If we need to update jedi-bundle (skylab4) in container, we need to extend coordination. Note that hard-wiring approach in cmakelist is not recommended.

**Comment by rhaesung** [comment]
> @rhaesung are you testing under singularity? If we need to update jedi-bundle (skylab4) in container, we need to extend coordination. Note that hard-wiring approach in cmakelist is not recommended.

@jkbk2004 No, I am testing on my feature branch. And note that the jedi-bundle (skylab4) is not tested yet on hera.internal baseline.

Thanks, the current modified cmakelist is just for testing purposes.

**Comment by rhaesung** [comment]
@jkbk2004 When running vector2tile with cmake generated executable, I get an error about illegal characters. The previous run without using cmake was successful and the loaded versions of NetCDF are both 4.7.0. Any idea? 

> Writing tile file: ./20191221.000000.sfc_data.tile1.nc
 NetCDF: Name contains illegal characters
vec2tile failed

**Comment by rhaesung** [comment]
@jkbk2004 Okay, it looks like we need to update the modulefiles/landda_hera.intel.lua correctly loading these modules: intel/2022.2, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4.

**Comment by rhaesung** [comment]
Here is the test folder I am seeing the below error message after following these steps:

> + /apps/oneapi/mpi/2021.6.0/bin/mpiexec -n 6 /scratch1/NCEPDEV/stmp2/Rhaesung.Kim/rhaesung/sync/land-DA_workflow_working/build/bin/apply_incr.exe /scratch1/NCEPDEV/stmp2/Rhaesung.Kim/rhaesung/sync/landda_expts/DA_GHCN_test1/DA/logs//apply_incr.log
/scratch1/NCEPDEV/stmp2/Rhaesung.Kim/rhaesung/sync/land-DA_workflow_working/build/bin/apply_incr.exe: error while loading shared libraries: libnetcdff.so.7: cannot open shared object file: No such file or directory

`module use modulefiles`
 `module load landda_hera.intel `
 `module load intel/2022.2.0 impi/2022.2.0 netcdf/4.7.0 netcdf-hdf5parallel/4.7.4 `
 `mkdir build `
 `cd build/ `
 `ecbuild .. `
 `make -j4 `
 `cd .. `
 `./do_submit_cycle.sh settings_DA_cycle_era5 `

**Comment by zach1221** [comment]
Hi, @natalie-perlin . In speaking with @rhaesung I think we will need to rebuild the spack-stack library for the land DA system. Eventually, we would need to have versions of fv3_mods_hera and ioda_mods_hera existing on all tier-1 machines, however we'd ideally like to begin testing with a new spack-stack library on Hera, that is built with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Would you be able to assist with this?

**Comment by natalie-perlin** [comment]
@zach1221 @yichengt90 @ulmononian 
  Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.

**Comment by zach1221** [comment]
> @zach1221 @yichengt90 @ulmononian Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.

Hey, @natalie-perlin @ulmononian following up here. Is this something you'd be able to help us with? Ideally, we'd like to merge this PR by end of next week, but we need a Hera spack installation with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Is this timeline realistic? @rhaesung feel free to chime in if I'm missing any pertinent info.

**Comment by rhaesung** [comment]
> > @zach1221 @yichengt90 @ulmononian Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.
> 
> Hey, @natalie-perlin @ulmononian following up here. Is this something you'd be able to help us with? Ideally, we'd like to merge this PR by end of next week, but we need a Hera spack installation with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Is this timeline realistic? @rhaesung feel free to chime in if I'm missing any pertinent info.

Thanks, @zach1221!
@natalie-perlin @clouden90 @ulmononian We also need these:

/scratch2/NCEPDEV/land/data/jedi/fv3_mods_hera
/scratch1/NCEPDEV/stmp2/Rhaesung.Kim/rhaesung/sync/land-DA_workflow/DA_update/ioda_mods_hera

**Comment by ulmononian** [comment]
> > @zach1221 @yichengt90 @ulmononian Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.
> 
> Hey, @natalie-perlin @ulmononian following up here. Is this something you'd be able to help us with? Ideally, we'd like to merge this PR by end of next week, but we need a Hera spack installation with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Is this timeline realistic? @rhaesung feel free to chime in if I'm missing any pertinent info.

would it be possible to upgrade the library versions used by landDA / this PR to utilize lib versions in the latest spack-stack release (1.4.1 on hera; https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-rdhpcs-hera). reverting to netcdf/4.7.0 is probably not something that can happen in any official spack-stack installation (for one, netcdf is split into netcdf-c and netcdf-fotran). also, if not, this could be handled in one of two ways:

1. the developers can maintain their own installation to accommodate the needs of this pr 
2. the spack-stack team can perform a test installation on a single machine (that will be deleted in ~10months). an issue on either the land DA or spack-stack repo with a **clear list** of the libraries needed should be submitted (please see https://github.com/JCSDA/spack-stack/blob/develop/project_charter.md#test-installations).

**Comment by ulmononian** [comment]
> > @zach1221 @yichengt90 @ulmononian Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.
> 
> Hey, @natalie-perlin @ulmononian following up here. Is this something you'd be able to help us with? Ideally, we'd like to merge this PR by end of next week, but we need a Hera spack installation with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Is this timeline realistic? @rhaesung feel free to chime in if I'm missing any pertinent info.

would it be possible to upgrade the library versions used by landDA / this PR to utilize lib versions in the latest spack-stack release (1.4.1 on hera; https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-rdhpcs-hera)? reverting to netcdf/4.7.0 is probably not something that can happen in any official spack-stack installation (for one, netcdf is split into netcdf-c and netcdf-fotran). if not, this could be handled in one of two ways:

1. the developers can maintain their own installation to accommodate the needs of this pr. 
2. the spack-stack team can perform a test installation on a single machine (that will be deleted in ~10months). an issue on either the land DA or spack-stack repo with a **clear list** of the libraries needed should be submitted (please see https://github.com/JCSDA/spack-stack/blob/develop/project_charter.md#test-installations).

please let us know which avenue sounds better.

@AlexanderRichert-NOAA please feel free to add anything i most definitely missed.

**Comment by rhaesung** [comment]
> > > @zach1221 @yichengt90 @ulmononian Let me also communicate with @ulmononian regarding this spack-stack installation, to clarify whether I'm taking over the spack-stack installations.
> > 
> > 
> > Hey, @natalie-perlin @ulmononian following up here. Is this something you'd be able to help us with? Ideally, we'd like to merge this PR by end of next week, but we need a Hera spack installation with intel/2022.2.0, impi/2022.2.0, netcdf/4.7.0, and netcdf-hdf5parallel/4.7.4. Is this timeline realistic? @rhaesung feel free to chime in if I'm missing any pertinent info.
> 
> would it be possible to upgrade the library versions used by landDA / this PR to utilize lib versions in the latest spack-stack release (1.4.1 on hera; https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html#noaa-rdhpcs-hera)? reverting to netcdf/4.7.0 is probably not something that can happen in any official spack-stack installation (for one, netcdf is split into netcdf-c and netcdf-fotran). if not, this could be handled in one of two ways:
> 
> 1. the developers can maintain their own installation to accommodate the needs of this pr.
> 2. the spack-stack team can perform a test installation on a single machine (that will be deleted in ~10months). an issue on either the land DA or spack-stack repo with a **clear list** of the libraries needed should be submitted (please see https://github.com/JCSDA/spack-stack/blob/develop/project_charter.md#test-installations).
> 
> please let us know which avenue sounds better.
> 
> @AlexanderRichert-NOAA please feel free to add anything i most definitely missed.

Thanks @ulmononian! How about netcdf-hdf5parallel/4.7.4? Looks like at least this module is required in the current code base.

**Comment by rhaesung** [comment]
@jkbk2004 @zach1221 Okay, at least on Hera, I was able to use the current spack version and update the libraries in landda_hera.intel.lua file. However, I got the below errors when loading the landda_orion.intel. @ulmononian Can you please take a look?

```
Lmod has detected the following error:  The following module(s) are unknown:
"gftl-shared/v1.5.0" "netcdf-c/4.9.0"
```

**Comment by ulmononian** [comment]
@rhaesung i don't understand. the hera lua file that you updated yesterday still points to the old stack (https://github.com/rhaesung/land-DA_workflow/blob/feature/ERA5baseline_sync/modulefiles/landda_hera.intel.lua). 

can you share the lua file you are trying to use for orion? the ones in this repo branch (why are there 2 for orion and 2 for hera?) are pretty outdated.

**Comment by rhaesung** [comment]
> @rhaesung i don't understand. the hera lua file that you updated yesterday still points to the old stack (https://github.com/rhaesung/land-DA_workflow/blob/feature/ERA5baseline_sync/modulefiles/landda_hera.intel.lua).
> 
> can you share the lua file you are trying to use for orion? the ones in this repo branch (why are there 2 for orion and 2 for hera?) are pretty outdated.

I've tested and confirmed that no stack update is needed for this pr. I just updated the current lua file (https://github.com/rhaesung/land-DA_workflow/blob/feature/ERA5baseline_sync/modulefiles/landda_hera.intel.lua). 

On orion, this one (https://github.com/rhaesung/land-DA_workflow/blob/feature/ERA5baseline_sync/modulefiles/landda_orion.intel.lua) worked fine until last week.

**Comment by ulmononian** [comment]
@rhaesung i was just able load the orion lua file you linked to on orion. i didn't get any lmod error for gftl-shared or netcdf-c. what exactly isn't working?

Screen Shot 2023-08-16 at 11 52 48 AM

**Comment by zach1221** [comment]
I can't load the module either for some reason, on orion. Maybe file system issue. Not sure. Strange Cameron can load it though.
image

**Comment by rhaesung** [comment]
> @rhaesung i was just able load the orion lua file you linked to on orion. i didn't get any lmod error for gftl-shared or netcdf-c. what exactly isn't working?
> 
> Screen Shot 2023-08-16 at 11 52 48 AM

@ulmononian it's weird. @zach1221 and I still got this error:

Screenshot 2023-08-16 at 2 59 28 PM

**Comment by zach1221** [comment]
I'm going to try loading the module via my pipeline on Orion. Might be user specific, system related issue.

**Comment by ulmononian** [comment]
can either of you access `/work/noaa/epic-ps/role-epic-ps`? this space was supposed to go away, so maybe weird permission issues.

**Comment by zach1221** [comment]
> can either of you access `/work/noaa/epic-ps/role-epic-ps`? this space was supposed to go away, so maybe weird permission issues.

Yah, I can cd to it and view contents.

**Comment by rhaesung** [comment]
> > can either of you access `/work/noaa/epic-ps/role-epic-ps`? this space was supposed to go away, so maybe weird permission issues.
> 
> Yah, I can cd to it and view contents.

okay, now I can load modules after replacing to:

> setenv("EPICHOME", "/work/noaa/epic/role-epic")
setenv("JEDI_INSTALL", pathJoin(os.getenv("EPICHOME"),"contrib/orion/v1.1"))

@ulmononian Can I load 'netcdf/4.7.0' and 'netcdf-hdf5parallel/4.7.4' using this lua file? If not, what's the other possible option?

**Comment by rhaesung** [comment]
@jkbk2004 @zach1221 ctest is now functional. Below is the instruction to run test on Hera:

> $ git clone -b feature/ERA5baseline_sync --recursive https://github.com/rhaesung/land-DA_workflow.git
$ ln -fs /scratch2/NAGAPE/epic/UFS_Land-DA/inputs
$ cd land-DA_workflow
$ module use modulefiles && module load landda_hera.intel
$ mkdir build && cd build
$ ecbuild ../ && make -j4
$ salloc --qos=debug --tasks-per-node=6 --account=epic --time=00:30:00
$ module use ../modulefiles && module load landda_hera.intel
$ ctest

**Comment by ulmononian** [comment]
> 'netcdf-hdf5parallel

cool! as far as i know, those modules are not available as orion system defaults and they are also not available in any of the spack-stack versions installed on orion. i am not sure exactly what `netcdf-hdf5parallel` is, but we don't currently build it in the spack-stack. i saw it was a sys. installed module on hera, but i am not sure if it is simply netcdf compiled against hdf5+mpi or what.

**Comment by jkbk2004** [comment]
@rhaesung Please, don't mix up system default modules with spack stack modules. All stack modules should come from spack stack. can you make sure about that?

File: modulefiles/landda_hera.intel.lua [hunk 1]
```diff
@@ -76,6 +76,7 @@ load(pathJoin("mapl", mapl_ver))
 
 load("ufs-pyenv")
 load("atlas")
+load("netcdf-hdf5parallel/4.7.4")
```
Comment (jkbk2004) [inline modulefiles/landda_hera.intel.lua]: @ulmononian can you review this change to confirm the package within spack stack?

Comment (ulmononian) [reply modulefiles/landda_hera.intel.lua]: this package is not provided as part of the spack-stack. i checked and it looks like a system default. i would not recommend using it.

**Comment by jkbk2004** [comment]
@ulmononian one question about modulefiles, why do I see 63) netcdf-hdf5parallel/4.7.4 ?
```
[role.epic@Hera:/scratch1/NCEPDEV/nems/role.epic/testing/landda/land-DA_workflow]# module load landda_hera.intel
[role.epic@Hera:/scratch1/NCEPDEV/nems/role.epic/testing/landda/land-DA_workflow]# module list

Currently Loaded Modules:
  1) intel/2022.1.2                   23) bacio/2.4.1                45) py-bottleneck/1.3.5
  2) stack-intel/2021.5.0             24) crtm-fix/2.4.0_emc         46) py-pyparsing/3.0.9
  3) impi/2022.1.2                    25) git-lfs/2.10.0             47) py-packaging/21.3
  4) stack-intel-oneapi-mpi/2021.5.1  26) crtm/2.4.0                 48) py-numexpr/2.8.3
  5) miniconda/3.9.12                 27) g2/3.4.5                   49) py-six/1.16.0
  6) stack-python/3.9.12              28) g2tmpl/1.10.2              50) py-python-dateutil/2.8.2
  7) cmake/3.23.1                     29) ip/3.3.3                   51) py-pytz/2022.2.1
  8) ecbuild/3.6.5                    30) sp/2.3.3                   52) py-pandas/1.4.0
  9) libjpeg/2.1.0                    31) w3emc/2.9.2                53) py-pyyaml/6.0
 10) jasper/2.0.32                    32) gftl/1.8.1                 54) ufs-pyenv/1.0.0
 11) zlib/1.2.13                      33) gftl-shared/1.5.0          55) boost/1.78.0
 12) libpng/1.6.37                    34) yafyaml/0.5.1              56) eigen/3.4.0
 13) pkg-config/0.27.1                35) mapl/2.22.0-esmf-8.3.0b09  57) eckit/1.20.2
 14) hdf5/1.14.0                      36) openblas/0.3.19            58) fftw/3.3.10
 15) curl/7.29.0                      37) py-setuptools/59.4.0       59) fckit/0.10.0
 16) zstd/1.5.2                       38) py-numpy/1.22.3            60) fiat/1.1.0
 17) netcdf-c/4.9.2                   39) py-cftime/1.0.3.4          61) ectrans/1.2.0
 18) netcdf-fortran/4.6.0             40) py-cython/0.29.32          62) atlas/0.32.1
 19) parallel-netcdf/1.12.2           41) py-f90nml/1.4.3            63) netcdf-hdf5parallel/4.7.4
 20) parallelio/2.5.9                 42) py-markupsafe/2.1.1        64) landda_hera.intel
 21) esmf/8.3.0b09                    43) py-jinja2/3.1.2
 22) fms/2022.04                      44) py-netcdf4/1.5.3
```

**Comment by ulmononian** [comment]
wasn't that just added to the lua file in a commit yesterday?

**Comment by jkbk2004** [comment]
> wasn't that just added to the lua file in a commit yesterday?

I see its system modulefile 
```
--------------------------------------- /apps/modules/modulefamilies/intel_impi ----------------------------------------
   hdf5parallel/1.10.5        itac/latest        netcdf-hdf5parallel/4.7.0        pnetcdf/1.7.0     pnetcdf/1.11.2 (D)
   hdf5parallel/1.10.6 (D)    itac/2019   (D)    netcdf-hdf5parallel/4.7.4 (D)    pnetcdf/1.10.0    tau/2.29
```

**Comment by jkbk2004** [comment]
@rhaesung why do you need to mix use system compiler and hdf modulefiles with spack stack?

**Comment by jkbk2004** [comment]
@rhaesung https://github.com/NOAA-EMC/ufs-land-driver it did not contain b1bfaf1efbc77f86bef0c2d0e88e94867cbe9033

**Comment by rhaesung** [comment]
> @rhaesung https://github.com/NOAA-EMC/ufs-land-driver it did not contain b1bfaf1efbc77f86bef0c2d0e88e94867cbe9033

@jkbk2004 @zach1221 just updated the ufs-land-driver submodule path

**Comment by rhaesung** [comment]
> @rhaesung why do you need to mix use system compiler and hdf modulefiles with spack stack?

@jkbk2004 I use spack stack compiler and just added ‘netcdf-hdf5parallel/4.7.4’ to the lua file.

To build the land da system, the current code base from PSL and EMC is required to load ‘netcdf/4.7.0’ and ‘netcdf-hdf5parallel/4.7.4’. On hera, netcdf/4.7.0 is loaded as system default so I don’t need to load it from the lua file.

**Comment by rhaesung** [comment]
@jkbk2004 @zach1221 the module issue is resolved. you can start testing

**Comment by jkbk2004** [comment]
@rhaesung land-DA pr was merged. Can you update hash and revert change in gitmodules. correct hash is  https://github.com/ufs-community/land-DA/commit/fabdfeb5dc9e07d1f008d57b6198c460995aa4f6

**Comment by jkbk2004** [comment]
@rhaesung hash is not corrcet. Correct one is https://github.com/ufs-community/land-DA/commit/fabdfeb5dc9e07d1f008d57b6198c460995aa4f6

**Comment by rhaesung** [comment]
Thanks, @jkbk2004 !

**Review by zach1221 — APPROVED** [review]

**Comment by rhaesung** [comment]
> @rhaesung hash is not corrcet. Correct one is [ufs-community/land-DA@fabdfeb](https://github.com/ufs-community/land-DA/commit/fabdfeb5dc9e07d1f008d57b6198c460995aa4f6)

@jkbk2004 it is resolved.

**Pull request merged**